Signal/src/network/GiphyDownloader.swift

```diff
@@ -0,0 +1,410 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import ObjectiveC
+
+// Stills should be loaded before full GIFs.
+enum GiphyRequestPriority {
+    case low, high
+}
+
+// Represents a request to download a GIF.
+//
+// Should be cancelled if no longer necessary.
+@objc class GiphyAssetRequest: NSObject {
```

Comment:
TIOLI: Consider using PromiseKit to drive the success, failure, cancel behavior, since it seems parallel.

Reply:
Let's discuss how you see that improving this code.  Note that the success and failure callbacks here are deliberately cleared in the cancel case and also in the completion case to avoid double-calling, etc.